### PR TITLE
Added Webhook support

### PR DESCRIPTION
**Pull request by AggressivelyMeows**

### CODE REVIEW

File: discord/webhook.py [hunk 1]
```diff
@@ -1,28 +1,32 @@
...
-        self.avatar = kwargs.get('avatar','')
-        self.name = kwargs.get('name','')
+AVATAR_BASE = 'https://cdn.discordapp.com/avatars/{id}/{avatar}.png'
+class Webhook(Hashable):
+    __slots__ = ( 'name', 'id', 'avatar', 'url', '_state',
+                 'webhook', 'token')
+    def __init__(self,state, **kwargs):
+        self._state = state
+        
+        self.name = kwargs.get('name',None)
```
Comment (TheTrain2000) [inline discord/webhook.py]: `None` is unnecessary, as it will be automatically returned if the key is not found.

File: discord/channel.py [hunk 2]
```diff
@@ -324,7 +326,59 @@ def is_me(m):
...
+        -------
+        Forbidden
+            You do not have proper permissions to do the actions required.
+        HTTPException
+            Getting webhooks failed
+
 
+        Returns
+        --------
+        List
```
Comment (Rapptz) [inline discord/channel.py]: This is List[:class:\`Webhook\`]

File: discord/channel.py [hunk 3]
```diff
@@ -324,7 +326,59 @@ def is_me(m):
...
+        webhooks = yield from self._state.http.get_channel_webhooks(self.id)
+        return [Webhook._from_data(self._state, wb) for wb in webhooks]
+    @asyncio.coroutine
+    def create_webhook(self,name,avatar):
+        """|coro|
+        Parameters
+        -----------
+        name: str
+            The name of the webhook
+        avatar: url or str
```
Comment (Rapptz) [inline discord/channel.py]: This type hint seems completely wrong.

File: discord/channel.py [hunk 4]
```diff
@@ -324,7 +326,59 @@ def is_me(m):
...
+        Raises
+        -------
+        Forbidden
+            You do not have proper permissions to do the actions required.
+        HTTPException
+            Creating the webhook failed.
+
+        Returns
+        --------
+        discord.Webhook - The created webhook
```
Comment (Rapptz) [inline discord/channel.py]: This is 

```
Returns
-----------
:class:`Webhook`
    The created webhook.
````

File: discord/channel.py [hunk 5]
```diff
@@ -324,7 +326,59 @@ def is_me(m):
...
+        """
+        if isinstance(avatar,str):
+            ret = yield from self._state.http._session.get(avatar)
+            if ret.status not in [200,204]:
+                m = yield from ret.text()
+                raise discord.HTTPException(ret,message=m)
+            data = yield from ret.read()
+            img = io.BytesIO(data)
+            img.seek(0)
+            avatar = utils._bytes_to_base64_data(img.read())
```
Comment (Rapptz) [inline discord/channel.py]: This entire segment probably belongs inside `HTTPClient`.

Or completely removed because nothing in the library allows setting a URL for an avatar/icon. Not to mention that this will probably always be invoked since the type hint says "URL or str" and you separate them as `isinstance(avatar, str)`, which will pretty much always be true.

File: discord/guild.py [hunk 6]
```diff
@@ -1153,3 +1153,14 @@ def audit_logs(self, *, limit=100, before=None, after=None, reverse=None, user=N
 
         return AuditLogIterator(self, before=before, after=after, limit=limit,
                                 reverse=reverse, user_id=user, action_type=action)
+    @asyncio.coroutine
+    def webhooks(self):
+        """
+        Gets all guild webhooks
+
+        Returns:
+            list:
```
Comment (Rapptz) [inline discord/guild.py]: This documentation is completely wrong.

It should be:

```
Returns
---------
List[:class:`Webhook`]
    The webhooks for the guild.
```

Also missing the `Raises` section.

File: discord/http.py [hunk 7]
```diff
@@ -524,6 +524,13 @@ def create_channel(self, guild_id, name, channe_type, permission_overwrites=None
     def delete_channel(self, channel_id, *, reason=None):
         return self.request(Route('DELETE', '/channels/{channel_id}', channel_id=channel_id), reason=reason)
 
+    def get_channel_webhooks(self, channel_id):
+        return self.request(Route('GET','/channels/{channel_id}/webhooks', channel_id=channel_id))
+    
+    def create_webhook(self, channel_id, *, name='', avatar=''):
```
Comment (Rapptz) [inline discord/http.py]: Are you sure these are appropriate default parameters?

File: discord/webhook.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+import aiohttp
+import io
+import json
+import asyncio
+
+# Webhook code made by Cerulean#7014 
+
+WEBHOOK_BASE = 'https://discordapp.com/api/webhooks/'
+AVATAR_BASE = 'https://cdn.discordapp.com/avatars/{id}/{avatar}.png'
+class Webhook(Hashable):
```
Comment (Rapptz) [inline discord/webhook.py]: This is missing documentation.

File: discord/webhook.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    __slots__ = ( 'name', 'id', 'avatar', 'url', '_state',
+                 'webhook', 'token')
+    def __init__(self,state, **kwargs):
+        self._state = state
+        
+        self.name = kwargs.get('name',None)
+        self.id = kwargs.get('id',0)
+        self.token = kwargs.get('token','')
+        self.url = WEBHOOK_BASE + '{0.id}/{0.token}'.format(self)
+        self.avatar = AVATAR_BASE.format(id=self.id,
```
Comment (Rapptz) [inline discord/webhook.py]: Nowhere in the library is `avatar` used to refer to the CDN. This is reserved for `avatar_url`.

File: discord/webhook.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
+from .embeds import Embed
+from .mixins import Hashable
+from .errors import HTTPException
+from .utils import _bytes_to_base64_data
+import aiohttp
+import io
+import json
+import asyncio
+
+# Webhook code made by Cerulean#7014
```
Comment (Rapptz) [inline discord/webhook.py]: Why is this here?

File: discord/webhook.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+from .errors import HTTPException
+from .utils import _bytes_to_base64_data
+import aiohttp
+import io
+import json
+import asyncio
+
+# Webhook code made by Cerulean#7014 
+
+WEBHOOK_BASE = 'https://discordapp.com/api/webhooks/'
```
Comment (Rapptz) [inline discord/webhook.py]: Are you sure this belongs here?

File: discord/webhook.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+AVATAR_BASE = 'https://cdn.discordapp.com/avatars/{id}/{avatar}.png'
```
Comment (Rapptz) [inline discord/webhook.py]: Why is this here? Just do it inline if you're going to use it once.

File: discord/webhook.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    def __init__(self,state, **kwargs):
+        self._state = state
+        
+        self.name = kwargs.get('name',None)
+        self.id = kwargs.get('id',0)
+        self.token = kwargs.get('token','')
+        self.url = WEBHOOK_BASE + '{0.id}/{0.token}'.format(self)
+        self.avatar = AVATAR_BASE.format(id=self.id,
+                                         avatar=kwargs.get('avatar',''))
+        self.webhook = kwargs # Anything we didnt get will be in here
```
Comment (Rapptz) [inline discord/webhook.py]: Why is this here?

Nowhere in the library is this done.

File: discord/webhook.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    @classmethod
+    def _from_data(cls,state,data):
+        return cls(state,**data)
```
Comment (Rapptz) [inline discord/webhook.py]: This is useless. Just use the constructor.

File: discord/webhook.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    @asyncio.coroutine
+    def edit(self,*,update=False,**kwargs):
+        """
+        Set the webhooks information
+
+        Returns
+        ------
+            Updated discord.Webhook
```
Comment (Rapptz) [inline discord/webhook.py]: Why is `edit` returning anything? Let this be an in-place edit.

File: discord/webhook.py [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+        self.avatar = AVATAR_BASE.format(id=self.id,
+                                         avatar=kwargs.get('avatar',''))
+        self.webhook = kwargs # Anything we didnt get will be in here
+    @classmethod
+    def _from_data(cls,state,data):
+        return cls(state,**data)
+    @asyncio.coroutine
+    def edit(self,*,update=False,**kwargs):
+        """
+        Set the webhooks information
```
Comment (Rapptz) [inline discord/webhook.py]: This documentation is woefully incomplete.

It doesn't detail any exceptions thrown or parameters accepted.

File: discord/webhook.py [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+
+        Returns
+        ------
+            Updated discord.Webhook
+        """
+        if kwargs.get('name'): self.name = kwargs.get('name')
+        if kwargs.get('avatar'): self.avatar = kwargs.get('avatar')
+        if update: yield from self.update() # Updates the webhook
```
Comment (Rapptz) [inline discord/webhook.py]: Nowhere in the library is this "idiom" done.

File: discord/webhook.py [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+        return self
```
Comment (Rapptz) [inline discord/webhook.py]: Don't return anything here.

File: discord/webhook.py [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+    @asyncio.coroutine
+    def update(self,**kwargs):
+        """|coro|
+        Updates the webhook
+        """
+        avatar = self.avatar
+        if self.avatar:
+            ret = yield from self._state.http._session.request('GET',self.avatar)
```
Comment (Rapptz) [inline discord/webhook.py]: Writing this should have triggered that something must have gone wrong in the implementation.

File: discord/webhook.py [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+    @asyncio.coroutine
+    def make_request(self,method,data):
+        """|coro|
+        Send data to discord
+
+        Has to use a different system than discord.py because of how different webhooks
+        are to the main client.
+
+        Danny, if you dont like how I did this, feel free to change it. I feel like I could
+        Have done better TBH
```
Comment (Rapptz) [inline discord/webhook.py]: 🙏

File: discord/webhook.py [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+        Sends a message to the webhook
+
+        Parameters:
+
+            message: str
+                The message you want to send
+            embeds: list [discord.Embed] or discord.Embed
+                The embed(s) you want to send
+            (Everything else is the same as the standard send_message)
```
Comment (Rapptz) [inline discord/webhook.py]: This documentation needs work.

**Comment by Rapptz** [comment]
I implemented this myself.

Thanks for the PR though.
